### DISCUSSION TITLE
Wrong Dependency Repository Link

### DISCUSSION THREAD

**Discussion by mikepenz** [community/community#19492] [Code Security]
Using the new [Dependency Submission API](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/using-the-dependency-submission-api) for `maven` packages links to a wrong repository. 

<img width="945" alt="Screenshot 2022-06-26 at 17 14 15" src="https://user-images.githubusercontent.com/1476232/175821012-b24629db-10bf-4f9c-be0e-f278042e4c37.png">

Found https://github.com/mikepenz/gradle-dependency-submission/network/dependencies
The original repository would be here: https://github.com/Kotlin/kotlinx.coroutines

Looking up this dependency on Maven (and it's pom.xml file) shows the right repository link: https://search.maven.org/artifact/org.jetbrains.kotlinx/kotlinx-coroutines-core/1.6.3/pom

Is there more information available how this mapping is created? 
I found some repositories on GitHub which map correctly to their repository. For example `OkHttp`, and `Okio` seen [here](https://github.com/mikepenz/MaterialDrawer/network/dependencies) 

<img width="941" alt="Screenshot 2022-06-26 at 17 17 20" src="https://user-images.githubusercontent.com/1476232/175821124-d6cca99b-9576-4360-8313-c4718e32a290.png">

Not sure if it is related, but for `OkHttp` it shows [Dependents](https://github.com/square/okhttp/network/dependents), but I can't find any information how this is done. Is it a manual claim process so a maven package gets linked with the respective repository?

**Comment by laughedelic** [comment]
I observed the same issue in a Scala repository (also submitting Maven dependencies): https://github.com/scalameta/metals/network/dependencies

![Screenshot 2022-07-05 at 01 40 54](https://user-images.githubusercontent.com/766656/177225734-9cf16d15-b0b7-4129-a388-be5403be6753.png)

- most of the dependencies are not linked at all
- `org.scala-lang:scala-compiler` is correctly linked to https://github.com/scala/scala
- `org.scala-lang:scala-library` is supposed to be linked to the same https://github.com/scala/scala repository, but is linked to an old fork instead: https://github.com/javagems/scala-library

Pom files for these artifacts contain correct `scm.url` information, e.g. https://search.maven.org/artifact/org.scala-lang/scala-library/2.13.8/pom

**Comment by larsgrefer** [comment]
I noticed the same problem here: https://github.com/freefair/gradle-plugins/network/dependencies

`org.springframework.boot:spring-boot-starter-json` is wrongly mapped to https://github.com/ling185831070/Spring-boot-demo instead of https://github.com/spring-projects/spring-boot

Aside from `pkg:maven` package URLs it even fails to correctly link to `pkg:github/gradle/gradle@a641d1caebe02ea109a69ca5fc0f59c20bceb9de`

**Comment by ckipp01** [comment]
Also noticing this same problem when submitting projects from the Mill build tool via https://github.com/ckipp01/mill-dependency-submission. You can see for example this being used in https://github.com/com-lihaoyi/mill/network/dependencies with many of the same issues listed above.

**Comment by lefou** [comment]
It's also worth mentioning, that many links don't link to projects at all. This is especially true for almost all Scala dependencies. At least some of them have correct Maven metadata in their published `pom.xml` files.

**Comment by courtneycl** [comment]
👋 Hi from the dependency graph team at GitHub -- just posting to say we see this and are digging into it. Thank you for the detailed comments!

**Reply by ckipp01** [reply]
Hey, just checking to see if there are any updates on this front and if there is anything the community could do to help with this?

**Reply by ckipp01** [reply]
Seeing that this has been a couple months and there has been no change. Is there any insight you can provide about how the linking is actually done? Again, is there anything the community can do to help here?

**Reply by ckipp01** [reply]
Pinging this again. All these issues still exist. Any updates or insight on the progress on this? Anything that we can follow?

**Reply by laughedelic** [reply]
@mikepenz @jhutchings1 could you unmark this as an answer? I think it might have gotten off the radar because it was marked as "solved".

**Reply by courtneycl** [reply]
Great apologies for the delay here; thank you for your continued patience. Package to repository mappings are fairly complicated; it uses a heuristic to identify the mappings since most package managers don't have authoritative mapping published at this time. Sometimes the heuristic gets it wrong and we'll do a manual override, which I've done for the packages listed in this issue and others I found. There are indeed a bunch where heuristic hasn't identified a mapping, particularly Scala dependencies. 

We have some ongoing work to replace our heuristics with using the registry as the authoritative source, and pulling from there when we can. We're doing this ecosystem by ecosystem and we haven't switched Maven over yet. Until we do that, we can manually remap them. I know it's not an ideal workflow. Feel free to add to this thread and ping me if you need one remapped. 

Reassigned: 
* `org.scala-lang:scala-library`
* `org.jetbrains.kotlinx:kotlinx-coroutines-*`
* `org.springframework.boot:spring-boot-starter-*`
* `org.springframework:spring-*`
* `com.beust:jcommander`
* `com.github.sbt:junit-interface`
* `com.google.javascript:closure-compiler-externs`
* `com.lmax:disruptor`
* `net.java.dev.jna:jna`
* `org.hamcrest:hamcrest-core`
* `org.mozilla:rhino`


Cleared (they were wrong; the source isn't listed in Maven Central):
* `com.google.code.findbugs:jsr305`
* `org.ow2.asm:asm-commons-*`

**Reply by ckipp01** [reply]
Thanks so much for the reply and update @courtneycl! I do have a couple questions if you'll indulge me:

> Package to repository mappings are fairly complicated; it uses a heuristic to identify the mappings since most package managers don't have authoritative mapping published at this time.

Could you explain a bit more about what you mean? For example when you look up these artifacts that you had listed under `Reassigned` they have valid POM files with `scm` entries that have the github url. For example just grabbing the first two:

- https://repo.maven.apache.org/maven2/org/scala-lang/scala-library/2.13.10/scala-library-2.13.10.pom
- https://repo.maven.apache.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines/0.19.2/kotlinx-coroutines-0.19.2.pom

I would assume that the heuristic is simply looking at these entries in the pom files? That would seem the most logical.

> Cleared (they were wrong; the source isn't listed in Maven Central):

I'm also not fully sure what you mean here. For example, here is the `jsr305` artifacts on Maven Central:
- https://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/

**Reply by larsgrefer** [reply]
For `pkg:maven`, the algorithm definitely should look for the corresponding pom file on maven central and check if there is a matching scm entry.

For `pkg:github` the matching should be trivial

**Reply by lefou** [reply]
Yeah, if Maven pom does not contain a usable `scm` entry and here is also no project `url` to use, it's probably better to not link at all.

**Reply by mikepenz** [reply]
Thanks @courtneycl 

In the small sample set I have there are for example more `okhttp` dependencies still referring to wrong repositories.
Looking at the `pom` file for these (like mentioned above) would point to the right github repo: https://repo1.maven.org/maven2/com/squareup/okhttp3/logging-interceptor/5.0.0-alpha.10/logging-interceptor-5.0.0-alpha.10.pom

<img width="876" alt="Screenshot 2022-11-24 at 09 22 07" src="https://user-images.githubusercontent.com/1476232/203730178-788b8758-3634-4cd5-ab3a-1220e495d943.png">

Also for the `org.jetbrains.kotlinx:kotlinx-coroutines-*` not all seem to be handled yet:
<img width="861" alt="Screenshot 2022-11-24 at 09 22 46" src="https://user-images.githubusercontent.com/1476232/203730291-ea96e51d-4f5c-401c-a602-42875ca90824.png">

**Reply by ckipp01** [reply]
Just found another incorrect one.

<img width="897" alt="Screenshot 2022-12-15 at 10 59 34" src="https://user-images.githubusercontent.com/13974112/207829934-cdb8669d-be18-4483-9361-9812d3bbfd73.png">

`bloop-config` 1.5.5 lives in it's own repo at https://github.com/scalacenter/bloop-config and the pom file that you can see [here](https://repo.maven.apache.org/maven2/ch/epfl/scala/bloop-config_2.13/1.5.5/bloop-config_2.13-1.5.5.pom) links to the correct repository. Currently this is just linking to itself.

**Reply by courtneycl** [reply]
@ckipp01 fixed! 

Coming back to your previous qs: 
> I would assume that the heuristic is simply looking at these entries in the pom files?

It looks at the pom files, but it also does other things, and sometimes the other things overwrite what we've detected from the pom files. It needs some rework for sure. 

> > Cleared (they were wrong; the source isn't listed in Maven Central):
> 
> I'm also not fully sure what you mean here. For example, here is the jsr305 artifacts on Maven Central

Sorry, I meant a GitHub-hosted source isn't listed. We currently only link packages to repos on GitHub.

**Reply by ckipp01** [reply]
So I don't really want to keep pointing these out as it seems like an incredibly inefficient way to do this, but looking at the deps in [Dotty](https://github.com/lampepfl/dotty/network/dependencies), there are a ton of them that are incorrect. Just to show a few:

<img width="895" alt="Screenshot 2023-02-01 at 16 41 38" src="https://user-images.githubusercontent.com/13974112/216090191-adcd03a8-98c3-4895-8df6-c9c3736e709f.png">
<img width="895" alt="Screenshot 2023-02-01 at 16 41 52" src="https://user-images.githubusercontent.com/13974112/216090200-6387a64a-183b-4436-915d-3a289e03153c.png">
<img width="895" alt="Screenshot 2023-02-01 at 16 40 14" src="https://user-images.githubusercontent.com/13974112/216090182-720aa904-6fb7-4cd0-be71-b43d37e6522e.png">

Seeing that this discussion was started 8 months ago, I'm more interested in any information you might have about a roadmap for better supporting various ecosystems on GitHub since it's been quite some time that the API has been open. For example even just ignoring the incorrect links even if you submit a maven dep with another build tool that's not Maven you still see weird messages like this:

<img width="1013" alt="Screenshot 2023-02-01 at 16 44 36" src="https://user-images.githubusercontent.com/13974112/216091303-e745b385-044f-4576-82ca-584376831ec3.png">

Showing that it can't find it in my `pom.xml` even though it clearly says right above it that it's coming from a `build.sbt`. Again, super happy this feature exists and I'm really looking forward to it being more usable from something besides pure java dependencies using poms. Is there any sort of public roadmap? 

Ah, I also just now am seeing https://github.com/orgs/community/discussions/43364. I would love to still do this if it's a possiblity.

**Comment by jhutchings1** [comment]
Appreciate all the feedback, folks! We need to do some major work on our package to repository mapping logic for some of these ecosystems. Please bear with us as we take some time to analyze and prioritize any needed changes.

**Reply by Tyler-Angell** [reply]
Thank you @jhutchings1 for looking into this, out of curiosity, are Security Advisories impacted by this? As a sample, `Microsoft.ApplicationInsights.WorkerService` in my repo is reporting the fork `vijay10193/asp1` as the source which hasn't been updated in a little over 3 years from the actual publisher `microsoft/ApplicationInsights-dotnet`. If the actual source repo publishes an advisory will I be notified of it without action from `vijay10193/asp1`

**Reply by jhutchings1** [reply]
No. We publish security advisories based on the package name/ecosystem, so that process sidesteps the package-to-repo mapping altogether. The package-to-repo mapping is purely used as a UI treatment in the dependency graph UX and we don't use that for any security conclusions.

**Reply by laughedelic** [reply]
@jhutchings1 it's been a while since the last update. Are there any plans or news you could share with us?

**Reply by courtneycl** [reply]
Hi @laughedelic -- we made some changes to our Maven repo mappings a ways back so that we'll only show mappings over a certain confidence level. This decreases the chance of having wrong mappings (though, not to zero). 

We are currently working on a bulk ingestion of metadata coming from ClearlyDefined -- this is primarily for licenses, but it also includes some repo mappings, and we'll be assessing whether we can use that as a source. Otherwise, we have time dedicated for mapping improvement efforts in the summer.

**Reply by laughedelic** [reply]
Thanks for the update! It's good to know that some work is happening or is planned. 

One concern about using metadata from any external source like ClearlyDefined is that it won't work for private repositories. As people in the threads above have mentioned, the metadata from the published `pom` files is probably the best source. 

Another possible approach could be extending the [Dependency Submission API](https://docs.github.com/en/rest/dependency-graph/dependency-submission?apiVersion=2022-11-28#create-a-snapshot-of-dependencies-for-a-repository) with the optional repository mapping. That would enable community tooling to solve the problem in the way that is best for each ecosystem.
